### PR TITLE
Add support for oneof

### PR DESCRIPTION
**Pull request by szakarias**

### CODE REVIEW

File: protoc_plugin/lib/message_generator.dart [hunk 1]
```diff
@@ -4,6 +4,19 @@
 
 part of protoc;
 
+class OneofEnumGenerator {
```
Comment (sigurdm) [inline protoc_plugin/lib/message_generator.dart]: Add a short dart-doc

File: protoc_plugin/lib/message_generator.dart [hunk 2]
```diff
@@ -4,6 +4,19 @@
...
+  static void generate(
+      IndentingWriter out, String classname, List<ProtobufField> fields) {
+    out.addBlock('enum ${classname} {', '}\n', () {
+      for (ProtobufField field in fields) {
+        final name = field.memberNames.fieldName;
+        out.println('$name, ');
+      }
+      out.println('unset');
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply protoc_plugin/lib/message_generator.dart]: According to my code-search skills we have no internal field named not_set or notSet.

File: protoc_plugin/lib/names.dart [hunk 3]
```diff
@@ -143,6 +172,13 @@ Iterable<String> defaultSuffixes() sync* {
   }
 }
 
+String oneofEnumClassName(
+    String descriptorName, Set<String> usedNames, String parent) {
```
Comment (sigurdm) [inline protoc_plugin/lib/names.dart]: parent => parentName

File: protoc_plugin/lib/names.dart [hunk 4]
```diff
@@ -200,10 +237,10 @@ Map<String, MemberNames> messageFieldNames(DescriptorProto descriptor,
     ..addAll(reservedMemberNames)
     ..addAll(reserved);
```
Comment (sigurdm) [inline protoc_plugin/lib/names.dart]: consider adding "notSet" here

File: protoc_plugin/pubspec.yaml [hunk 5]
```diff
@@ -10,7 +10,7 @@ environment:
 dependencies:
   fixnum: ^0.10.5
   path: ^1.0.0
-  protobuf: ^0.10.5
+  protobuf: ^0.10.6
```
Comment (sigurdm) [inline protoc_plugin/pubspec.yaml]: If this is a breaking change we should change the middle number.

File: protoc_plugin/test/protos/oneof.proto [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        bytes fourth = 7;
+        Bar fifth = 8;
+        enum_type sixth = 6;
+    }
+
+    oneof colors {
+        string red = 2;
+        string green = 4;
+    }
+}
```
Comment (sigurdm) [inline protoc_plugin/test/protos/oneof.proto]: Add newline

File: protoc_plugin/test/protos/oneof.proto [hunk 7]
```diff
@@ -0,0 +1,32 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+syntax = "proto3";
+
+package protobuf_unittest;
```
Comment (sigurdm) [inline protoc_plugin/test/protos/oneof.proto]: Could you also add a oneof somewhere in the _leading_underscore.proto?

File: protoc_plugin/CHANGELOG.md [hunk 8]
```diff
@@ -1,3 +1,8 @@
+## 0.13.0
+
+* Breaking change: Support for [oneof](https://developers.google.com/protocol-buffers/docs/proto3#maps)
```
[the remark this answers is not in the document]
Comment (szakarias) [reply protoc_plugin/CHANGELOG.md]: Done

File: protoc_plugin/CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,8 @@
...
+  Generated files require package:protobuf version 0.10.6 or newer.
+
```
[the remark this answers is not in the document]
Comment (szakarias) [reply protoc_plugin/CHANGELOG.md]: I agree, that was the plan.

File: protoc_plugin/lib/names.dart (hunk 3 above)
Comment (szakarias) [reply protoc_plugin/lib/names.dart]: Done

File: protoc_plugin/test/protos/oneof.proto (hunk 6 above)
Comment (szakarias) [reply protoc_plugin/test/protos/oneof.proto]: Done

File: protoc_plugin/CHANGELOG.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (szakarias) [reply protoc_plugin/CHANGELOG.md]: Done

File: protoc_plugin/lib/message_generator.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (szakarias) [reply protoc_plugin/lib/message_generator.dart]: Done.

File: protoc_plugin/test/protos/oneof.proto (hunk 7 above)
Comment (szakarias) [reply protoc_plugin/test/protos/oneof.proto]: No, since that one is `syntax = "proto2"`.

Comment (szakarias) [reply protoc_plugin/test/protos/oneof.proto]: Done

File: protobuf/lib/src/protobuf/field_set.dart [hunk 10]
```diff
@@ -48,6 +48,10 @@ class _FieldSet {
   /// Contains all the unknown fields, or null if there aren't any.
   UnknownFieldSet _unknownFields;
 
+  // Maps a oneof decl index to the tag number which is currently set. If the
+  // index is not present, the oneof field is unset.
+  Map<int, int> oneofCases = <int, int>{};
```
Comment (kevmoo) [inline protobuf/lib/src/protobuf/field_set.dart]: could be final?

File: protobuf/CHANGELOG.md [hunk 11]
```diff
@@ -1,3 +1,8 @@
+## 0.10.6
+
+* Added support for [oneof](https://developers.google.com/protocol-buffers/docs/proto3#oneof).
+
+
```
Comment (kevmoo) [inline protobuf/CHANGELOG.md]: nit: extra line

File: protobuf/lib/src/protobuf/field_set.dart (hunk 10 above)
Comment (szakarias) [reply protobuf/lib/src/protobuf/field_set.dart]: Done

File: protobuf/CHANGELOG.md (hunk 11 above)
Comment (szakarias) [reply protobuf/CHANGELOG.md]: done.

**Pull request merged**